### PR TITLE
refactor: track datastore query count in datastore wrapper

### PR DESCRIPTION
**Pull request by miparnisari**
## Description

**Problem**: Keeping track of `datastore_query_count` value within Check, ReverseExpand, ListObjects and ListUsers is tedious, bloats the code, breaks Single Responsibility Principle, and worse of all will mean breaking tests whenever we reduce the number of DB calls made.

**Proposal of this PR**: 

- keep track of `datastore_query_count` in a wrapper over the `OpenFGAStorage` object.
- `ResolutionMetadata.DatastoreQueryCount` is only filled at the end of the request.

⚠️ important files to review:
- `pkg/server/commands/check_command.go`
- `pkg/server/commands/list_objects.go`
- `pkg/server/commands/listusers/list_users_rpc.go`

⚠️ as a side effect of this change, Check API will change as follows: before, `datastore_query_count` was increased per path in the resolution tree. Now, it's increased globally. This means that we are now counting all the paths that we explored that didn't result in `allowed=true`.

## References

Closes https://github.com/openfga/openfga/issues/1477

### CODE REVIEW

File: internal/graph/check.go [hunk 1]
```diff
@@ -230,8 +226,7 @@ func union(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandle
 	return &ResolveCheckResponse{
 		Allowed: false,
 		ResolutionMetadata: &ResolveCheckResponseMetadata{
```
Comment (miparnisari) [inline internal/graph/check.go:228]: i want to send another PR where `ResolveCheckResponseMetadata` is not a pointer.

File: pkg/server/commands/list_objects.go [hunk 2]
```diff
@@ -270,11 +269,12 @@ func (q *ListObjectsQuery) evaluate(
 		reverseExpandResultsChan := make(chan *reverseexpand.ReverseExpandResult, 1)
 		objectsFound := atomic.Uint32{}
 
+		metricsDs := storagewrappers.NewMetricsOpenFGAStorage(q.datastore)
 		ds := storagewrappers.NewCombinedTupleReader(
-			q.datastore,
+			storagewrappers.NewBoundedConcurrencyTupleReader(
+				metricsDs, q.maxConcurrentReads),
 			req.GetContextualTuples().GetTupleKeys(),
 		)
```
Comment (miparnisari) [inline pkg/server/commands/list_objects.go:277]: FYI i dislike how this looks. Discussed offline with @jpadilla, the current problem we have is that some datastore wrappers are server-wide, while others are request-specific.

What we could do in a future PR is to have every query object create it's own datastore wrapper _from scratch_. The server should _only_ hold any cache objects that those wrappers need.

File: pkg/storage/storagewrappers/metrics.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	}
+}
+
+func (m *MetricsOpenFGAStorage) increase() {
+	m.counter.Add(1)
+}
+
+// Read see [storage.RelationshipTupleReader.ReadUserTuple].
+func (m *MetricsOpenFGAStorage) Read(ctx context.Context, store string, tupleKey *openfgav1.TupleKey, options storage.ReadOptions) (storage.TupleIterator, error) {
+	defer m.increase()
```
Comment (elbuo8) [inline pkg/storage/storagewrappers/metrics.go:42]: Why use `defer`? It's just an addition to the call stack with no point 🤔 More operations overall to process.

File: internal/mocks/mock_storage.go [hunk 4]
```diff
@@ -22,7 +22,6 @@ import (
 type MockTupleBackend struct {
 	ctrl     *gomock.Controller
 	recorder *MockTupleBackendMockRecorder
-	isgomock struct{}
```
[the remark this answers is not in the document]
Comment (jpadilla) [reply internal/mocks/mock_storage.go:25]: expected https://github.com/uber-go/mock/blob/main/CHANGELOG.md#050-15-oct-2024

**Review by jpadilla — DISMISSED** [review]

File: pkg/server/commands/check_command.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ type CheckQuery struct {
 	logger          logger.Logger
 	checkResolver   graph.CheckResolver
 	typesys         *typesystem.TypeSystem
-	datastore       storage.RelationshipTupleReader
+	datastore       *storagewrappers.MetricsOpenFGAStorage
```
Comment (elbuo8) [inline pkg/server/commands/check_command.go:33]: This shouldn't change at all.

File: pkg/server/commands/check_command.go [hunk 6]
```diff
@@ -74,7 +74,7 @@ func WithCacheController(ctrl cachecontroller.CacheController) CheckQueryOption
 func NewCheckCommand(datastore storage.RelationshipTupleReader, checkResolver graph.CheckResolver, typesys *typesystem.TypeSystem, opts ...CheckQueryOption) *CheckQuery {
 	cmd := &CheckQuery{
 		logger:             logger.NewNoopLogger(),
-		datastore:          datastore,
+		datastore:          storagewrappers.NewMetricsOpenFGAStorage(datastore),
```
Comment (elbuo8) [inline pkg/server/commands/check_command.go:77]: Store this in a var above, so the interface doesn't need to change, and subsequently have you have access to the additional methods you require.

File: pkg/storage/storagewrappers/metrics.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+// It is crucial that the wrapped object does NOT return results from an in-memory cache for this object to return accurate metrics.
+func NewMetricsOpenFGAStorage(wrapped storage.RelationshipTupleReader) *MetricsOpenFGAStorage {
+	return &MetricsOpenFGAStorage{
+		RelationshipTupleReader: wrapped,
+		counter:                 atomic.Uint32{},
+	}
+}
+
+type Metrics struct {
+	DatastoreQueryCount int
```
Comment (elbuo8) [inline pkg/storage/storagewrappers/metrics.go:30]: If you are casting it everywhere, why not use `uint32`?

File: pkg/server/commands/check_command.go (hunk 6 above)
Comment (miparnisari) [reply pkg/server/commands/check_command.go:77]: If i have the command object keep a reference to both `datastore` and `instrumentedDatastore`, then i wouldn't reference the former anywhere 🤔

**Review by elbuo8 — APPROVED** [review]

File: internal/graph/check.go (hunk 1 above)
Comment (miparnisari) [reply internal/graph/check.go:232]: https://github.com/openfga/openfga/pull/2061
